File: app/src/main/java/com/tcc/sicv/data/firebase/MaintenanceRepository.java
package com.tcc.sicv.data.firebase;

import android.support.annotation.NonNull;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;
import com.tcc.sicv.base.Result;
import com.tcc.sicv.data.model.Logs;
import com.tcc.sicv.data.model.MaintenanceVehicle;
import com.tcc.sicv.data.model.Vehicle;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Locale;

import static com.tcc.sicv.utils.Constants.CODE_VEHICLE_FIELD;
import static com.tcc.sicv.utils.Constants.COST_FIELD;
import static com.tcc.sicv.utils.Constants.DATE_FIELD;
import static com.tcc.sicv.utils.Constants.DATE_FORMAT;
import static com.tcc.sicv.utils.Constants.DESCRIPTION_FIELD;
import static com.tcc.sicv.utils.Constants.IMAGE_FIELD;
import static com.tcc.sicv.utils.Constants.LOGS_COLLECTION_PATH;
import static com.tcc.sicv.utils.Constants.MAINTENANCE_COLLECTION_PATH;
import static com.tcc.sicv.utils.Constants.MAINTENANCE_FIELD;
import static com.tcc.sicv.utils.Constants.MODEL_FIELD;
import static com.tcc.sicv.utils.Constants.RELEASE_VEHICLE_FIELD;
import static com.tcc.sicv.utils.Constants.USER_COLLECTION_PATH;
import static com.tcc.sicv.utils.Constants.VEHICLES_COLLECTION_PATH;

public class MaintenanceRepository {
    private final FirebaseFirestore db;

    public MaintenanceRepository() {
        db = FirebaseFirestore.getInstance();
    }

    public void setVehicleInMaintenance(
            final String email,
            Vehicle vehicle,
            final String date,
            final Result<MaintenanceVehicle> result
    ) {
        DocumentReference document = db.collection(USER_COLLECTION_PATH)
                .document(email)
                .collection(MAINTENANCE_COLLECTION_PATH)
                .document();

        final MaintenanceVehicle maintenanceVehicle = new MaintenanceVehicle(
                document.getId(), vehicle.getModelo(), vehicle.getImagem(),
                vehicle.getCodigo(), false
        );
        document
                .set(maintenanceVehicle)
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        result.onFailure(e);
                    }
                }).addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                        public void onSuccess(Void aVoid) {
                        setFirstLogInMaintenance(email, maintenanceVehicle, date, result);
                    }
                });
    }

    private void setFirstLogInMaintenance(
            final String email, final MaintenanceVehicle maintenanceVehicle,
            String date,
            final Result<MaintenanceVehicle> result
    ){
        String firstLogDescription = "Entrega do veículo a concessionária";
        db.collection(USER_COLLECTION_PATH).document(email).collection(MAINTENANCE_COLLECTION_PATH)
                .document(maintenanceVehicle.getMaintenanceCode())
                .collection(LOGS_COLLECTION_PATH)
                .add(new Logs(date, firstLogDescription, "R$ 0,00"))
                .addOnSuccessListener(new OnSuccessListener<DocumentReference>() {
                    @Override
                    public void onSuccess(DocumentReference documentReference) {
                        setMyVehicleInMaintenance(email, maintenanceVehicle, result);
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        result.onFailure(e);
                    }
                });
    }

    private void setMyVehicleInMaintenance(
            String email, final MaintenanceVehicle maintenance,
            final Result<MaintenanceVehicle> result
    ){
        db.collection(USER_COLLECTION_PATH).document(email).collection(VEHICLES_COLLECTION_PATH)
                .document(maintenance.getCod_veiculo())
                .update(MAINTENANCE_FIELD, true)
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {
                        result.onSuccess(maintenance);
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        result.onFailure(e);
                    }
                });
    }

    public void getVehiclesInMaintenance(
            final String email,
            final Result<ArrayList<MaintenanceVehicle>> result
    ) {
        db.collection(USER_COLLECTION_PATH).document(email).collection(MAINTENANCE_COLLECTION_PATH)
                .orderBy(MODEL_FIELD)
                .get()
                .addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
                    @Override
                    public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                        ArrayList<MaintenanceVehicle> maintenanceList = new ArrayList<>();
                        for (DocumentSnapshot item : queryDocumentSnapshots.getDocuments()) {
                            MaintenanceVehicle maintenance = new MaintenanceVehicle(
                                    item.getId(),
                                    (String) item.get(MODEL_FIELD),
                                    (String) item.get(IMAGE_FIELD),
                                    (String) item.get(CODE_VEHICLE_FIELD),
                                    item.getBoolean(RELEASE_VEHICLE_FIELD)
                            );
                            maintenanceList.add(maintenance);
                        }
                        result.onSuccess(maintenanceList);
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        result.onFailure(e);
                    }
                });
    }

    public void getLogsInMaintenance(
            String email,
            String maintenanceCode,
            final Result<ArrayList<Logs>> result
    ) {
        db.collection(USER_COLLECTION_PATH).document(email)
                .collection(MAINTENANCE_COLLECTION_PATH)
                .document(maintenanceCode)
                .collection(LOGS_COLLECTION_PATH)
                .get()
                .addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
                    @Override
                    public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                        ArrayList<Logs> logsList = new ArrayList<>();
                        for (DocumentSnapshot item : queryDocumentSnapshots.getDocuments()) {
                            logsList.add(new Logs(
                                    (String) item.get(DATE_FIELD),
                                    (String) item.get(DESCRIPTION_FIELD),
                                    (String) item.get(COST_FIELD)
                            ));
                        }
                        orderListByDate(logsList);
                        result.onSuccess(logsList);
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        result.onFailure(e);
                    }
                });
    }

    private void orderListByDate(ArrayList<Logs> logsList) {
        Collections.sort(logsList, new Comparator<Logs>() {
            DateFormat f = new SimpleDateFormat(DATE_FORMAT, Locale.getDefault());
            @Override
            public int compare(Logs o1, Logs o2) {
                try {
                    return f.parse(o1.getData()).compareTo(f.parse(o2.getData()));
                } catch (ParseException e) {
                    throw new IllegalArgumentException(e);
                }
            }
        });
    }
}

File: app/src/main/java/com/tcc/sicv/data/firebase/MaintenanceRepository.kt
package com.tcc.sicv.data.firebase

import com.google.firebase.firestore.FirebaseFirestore
import com.tcc.sicv.base.Result
import com.tcc.sicv.data.model.Logs
import com.tcc.sicv.data.model.MaintenanceVehicle
import com.tcc.sicv.data.model.Vehicle
import com.tcc.sicv.utils.Constants.CODE_VEHICLE_FIELD
import com.tcc.sicv.utils.Constants.COST_FIELD
import com.tcc.sicv.utils.Constants.DATE_FIELD
import com.tcc.sicv.utils.Constants.DESCRIPTION_FIELD
import com.tcc.sicv.utils.Constants.IMAGE_FIELD
import com.tcc.sicv.utils.Constants.LOGS_COLLECTION_PATH
import com.tcc.sicv.utils.Constants.MAINTENANCE_COLLECTION_PATH
import com.tcc.sicv.utils.Constants.MAINTENANCE_FIELD
import com.tcc.sicv.utils.Constants.MODEL_FIELD
import com.tcc.sicv.utils.Constants.RELEASE_VEHICLE_FIELD
import com.tcc.sicv.utils.Constants.USER_COLLECTION_PATH
import com.tcc.sicv.utils.Constants.VEHICLES_COLLECTION_PATH
import java.util.*

class MaintenanceRepository {
    private val db: FirebaseFirestore = FirebaseFirestore.getInstance()

    fun setVehicleInMaintenance(
            email: String,
            vehicle: Vehicle,
            date: String,
            result: Result<MaintenanceVehicle>
    ) {
        val document = db.collection(USER_COLLECTION_PATH)
                .document(email)
                .collection(MAINTENANCE_COLLECTION_PATH)
                .document()

        val maintenanceVehicle = MaintenanceVehicle(
                document.id, vehicle.modelo, vehicle.imagem,
                vehicle.codigo?:"", false
        )
        document
                .set(maintenanceVehicle)
                .addOnFailureListener { e -> result.onFailure(e) }.addOnSuccessListener { setFirstLogInMaintenance(email, maintenanceVehicle, date, result) }
    }

    private fun setFirstLogInMaintenance(
            email: String, maintenanceVehicle: MaintenanceVehicle,
            date: String,
            result: Result<MaintenanceVehicle>
    ) {
        val firstLogDescription = "Entrega do veículo a concessionária"
        db.collection(USER_COLLECTION_PATH).document(email).collection(MAINTENANCE_COLLECTION_PATH)
                .document(maintenanceVehicle.maintenanceCode)
                .collection(LOGS_COLLECTION_PATH)
                .add(Logs(date, firstLogDescription, "R$ 0,00"))
                .addOnSuccessListener { setMyVehicleInMaintenance(email, maintenanceVehicle, result) }
                .addOnFailureListener { e -> result.onFailure(e) }
    }

    private fun setMyVehicleInMaintenance(
            email: String, maintenance: MaintenanceVehicle,
            result: Result<MaintenanceVehicle>
    ) {
        db.collection(USER_COLLECTION_PATH).document(email).collection(VEHICLES_COLLECTION_PATH)
                .document(maintenance.cod_veiculo)
                .update(MAINTENANCE_FIELD, true)
                .addOnSuccessListener { result.onSuccess(maintenance) }
                .addOnFailureListener { e -> result.onFailure(e) }
    }

    fun getVehiclesInMaintenance(
            email: String,
            result: Result<MutableList<MaintenanceVehicle>>
    ) {
        db.collection(USER_COLLECTION_PATH).document(email).collection(MAINTENANCE_COLLECTION_PATH)
                .orderBy(MODEL_FIELD)
                .get()
                .addOnSuccessListener { queryDocumentSnapshots ->
                    val maintenanceList = ArrayList<MaintenanceVehicle>()
                    for (item in queryDocumentSnapshots.documents) {
                        val maintenance = MaintenanceVehicle(
                                item.id,
                                (item.get(MODEL_FIELD) as String?) ?:"",
                                (item.get(IMAGE_FIELD) as String?) ?:"",
                                (item.get(CODE_VEHICLE_FIELD) as String?) ?:"",
                                item.getBoolean(RELEASE_VEHICLE_FIELD)
                        )
                        maintenanceList.add(maintenance)
                    }
                    result.onSuccess(maintenanceList)
                }
                .addOnFailureListener { e -> result.onFailure(e) }
    }

    fun getLogsInMaintenance(
            email: String,
            maintenanceCode: String,
            result: Result<ArrayList<Logs>>
    ) {
        db.collection(USER_COLLECTION_PATH).document(email)
                .collection(MAINTENANCE_COLLECTION_PATH)
                .document(maintenanceCode)
                .collection(LOGS_COLLECTION_PATH)
                .get()
                .addOnSuccessListener { queryDocumentSnapshots ->
                    val logsList = ArrayList<Logs>()
                    for (item in queryDocumentSnapshots.documents) {
                        logsList.add(Logs(
                                (item.get(DATE_FIELD) as String?) ?:"",
                                (item.get(DESCRIPTION_FIELD) as String?) ?:"",
                                (item.get(COST_FIELD) as String?) ?:""
                        ))
                    }
                    result.onSuccess(logsList)
                }
                .addOnFailureListener { e -> result.onFailure(e) }
    }
}
